src/Common/src/System/Collections/Generic/SparseArrayBuilder.cs

```diff
@@ -0,0 +1,197 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Represents a reserved region within a <see cref="SparseArrayBuilder{T}"/>.
+    /// </summary>
+    [DebuggerDisplay("{DebuggerDisplay,nq}")]
+    internal struct Marker
+    {
+        /// <summary>
+        /// Constructs a new marker.
+        /// </summary>
+        /// <param name="count">The number of items to reserve.</param>
+        /// <param name="index">The index in the builder where this marker starts.</param>
+        public Marker(int count, int index)
+        {
+            Debug.Assert(count >= 0);
+            Debug.Assert(index >= 0);
+
+            Count = count;
+            Index = index;
+        }
+
+        /// <summary>
+        /// The number of items to reserve.
+        /// </summary>
+        public int Count { get; }
+
+        /// <summary>
+        /// The index in the builder where this marker starts.
+        /// </summary>
+        public int Index { get; }
+
+        private string DebuggerDisplay => $"{nameof(Index)}: {Index}, {nameof(Count)}: {Count}";
+    }
+
+    /// <summary>
+    /// Helper type for building arrays where sizes of certain segments are known in advance.
+    /// </summary>
+    /// <typeparam name="T">The element type.</typeparam>
+    internal struct SparseArrayBuilder<T>
+    {
+        /// <summary>
+        /// The underlying builder that stores items from non-reserved regions.
+        /// </summary>
+        /// <remarks>
+        /// This field is a mutable struct; do not mark it readonly.
+        /// </remarks>
+        private LargeArrayBuilder<T> _builder;
+
+        /// <summary>
+        /// The list of reserved regions within this builder.
+        /// </summary>
+        /// <remarks>
+        /// This field is a mutable struct; do not mark it readonly.
+        /// </remarks>
+        private ArrayBuilder<Marker> _markers;
+
+        /// <summary>
+        /// The total number of reserved slots within this builder.
+        /// </summary>
+        private int _reservedCount;
+
+        /// <summary>
+        /// Constructs a new builder.
+        /// </summary>
+        /// <param name="initialize">Pass <c>true</c>.</param>
+        public SparseArrayBuilder(bool initialize)
+        {
+            // Once C# gains parameterless struct constructors, please
+            // remove this workaround.
+            Debug.Assert(initialize);
+
+            _builder = new LargeArrayBuilder<T>(initialize: true);
```

Comment:
I haven't looked into it extensively yet, but I'm conjecturing that some of the speed-wise regressions in the tests for small collections could be abated if we could switch to the `: this()`-style pattern here and in `LargeArrayBuilder`. The JIT does not handle large structs very well yet, and additionally will not be able to inline either of these methods.